cloudpickle/cloudpickle.py

```diff
@@ -249,6 +252,93 @@ def _walk_global_ops(code):
                 yield op, instr.arg
 
 
+def _memoryview_from_bytes(data, format, readonly):
+    if not readonly:
+        # Forcibly render the recently unpickled bytes buffer writable:
+        # this is a violation of the immutability of a Python bytes object but
+        # this should be safe because:
+        # - data is a temporary variable that has just been allocated from
+        #   reading from the pickle stream and will never be used outside of
+        #   the scope of this reducer.
+        # - bytes objects are not subject to interning.
```

Comment:
Hum. I concluded this from the following:

```
>>> sys.intern(b'x')
Traceback (most recent call last):
  File "<ipython-input-5-02fdb271b026>", line 1, in <module>
    sys.intern(b'x')
TypeError: intern() argument 1 must be str, not bytes
```

Do you know the maximum length of interned bytes objects?

Reply:
Apparently (from experiments), only instances with zero or one-byte length are "interned" although I could not google that this is supposed to be officially the case. I am having a hard time understanding the C code of the `bytesobject.c` file.

Reply:
In pandas we just check the refcount of the object. Regardless of how interning happens, this will be safe if the refcount is exactly 1 AND you steal the reference. We just take a copy if the refcount is not 1. If the refcount is > 1 and the len is > 1, we raise a performance warning stating that something odd happened and a copy was done; however that doesn't happen. We don't warn under len 1 because a copy of 1 or two bytes is trivial.

Reply:
In cloudpickle we have to stay 100% pure python (hence this hack in ctypes). Furthermore we have to respect the pickle protocol and this function is called by the load function when hitting the REDUCE opcode. At this point the `bytes` buffer has already been read and is poped from the unpickler stack, but it's still in the lexical scope of the caller function (`load_reduce` IIRC).

However this bytes object will never be used anywhere else, so I am still quite confident that this ctypes hack, when called in the context of the unpickler is "safe".

Reply:
@ogrisel the language doesn't guarantee anything here, so you are on your own if you try to mutate a bytes object without taking the kind of precaution @llllllllll talks about.

Reply:
I am not saying that we need to use the C implementation shown in pandas, I am saying that we should copy the refcount checks that it does. We may also want to try harder to hide the mutable bytes object

Reply:
Thinking more about it, it might be possible to get the refcount down from 2 to 1 by nesting 2 reducers and doing the "mutate_move" in the wrapping reducer. I can also try to hide the `_base` reference with a readonly property.

Reply:
I have pushed a fix. It still lacks a more complete test harness though. Will work on that next week.